### PR TITLE
[ESIMD] Implement unified memory API - part1 - block_load(usm_ptr, ...)

### PR DESCRIPTION
**Pull request by v-klochkov**
The first patch implements block_load() API accepting USM pointers.
The new API uses optional compile-time cache-hint and alignment properties,
which makes it usable for PVC block loads that previously were implemented
using draft/experimental API lsc_block_load().

Some of the new block_load(usm, ) functions support the 'offset' argument,
which is not 'required' as the offset could be included into usm_ptr before
the block_load() call. That is done to make USM and ACCESSOR-based API
consistent (ACCESSOR-based API requires 'offset' parameter).

This patch is the 1st in a series of similar upcoming changes:
  block_load(accessor, ...)
  block_store(usm_ptr, ...),
  block_store(accessor, ...),
  gather()
  scatter()
  fence()
  atomic_update()

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 1]
```diff
@@ -581,6 +581,77 @@ ESIMD_INLINE __ESIMD_NS::SurfaceIndex __esimd_get_surface_index(MemObjTy obj)
...
+/// as a single \ref simd object.
+///
+/// @tparam Ty is element type.
+/// @tparam L1H is L1 cache hint.
+/// @tparam L3H is L3 cache hint.
+/// @tparam AddressScale is the address scale.
+/// @tparam ImmOffset is the immediate offset added to each address.
+/// @tparam DS is the data size.
+/// @tparam VS is the number of elements to load per address.
+/// @tparam Transposed indicates if the data is transposed during the transfer.
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:597]: Transposed here doesn't match actual name `_Transposed`

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 2]
```diff
@@ -581,6 +581,77 @@ ESIMD_INLINE __ESIMD_NS::SurfaceIndex __esimd_get_surface_index(MemObjTy obj)
...
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:635]: Same typo

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+
+/// simd<T, N> block_load(const T* ptr, simd_mask<1> pred,
+///                       simd<T, N> old_values, props={});             // (5)
+/// simd<T, N> block_load(const T* ptr, size_t offset, simd_mask<1> pred,
+///                       simd<T, N> old_values, props={});             // (6)
+
+/// simd<T, N> block_load(const T* ptr, props={}); // (1)
+/// This function loads a contiguous memory block from USM pointer \p ptr.
+///
+/// It may have temporary restrictions depending on L1, L2 cache hints,
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:574]:
```suggestion
/// There may be temporary restrictions depending on L1, L2 cache hints,
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+/// See details at 'Restrictions' section below. The restrictions will be
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:575]:
```suggestion
/// See details in the 'Restrictions' section below. The restrictions will be
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 5]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+/// relaxed eventually/soon.
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:576]:
```suggestion
/// relaxed in the future.
```

Maybe we don't commit to anything :)

File: sycl/test-e2e/ESIMD/esimd_test_utils.hpp [hunk 6]
```diff
@@ -8,8 +8,10 @@
 
 #pragma once
 
+#include <sycl/bit_cast.hpp>
 #include <sycl/ext/intel/esimd.hpp>
 #include <sycl/sycl.hpp>
+
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/esimd_test_utils.hpp:14]: Nit: Remove whitespace

File: sycl/test/esimd/memory_properties.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
+// RUN: %clangxx -O0 -fsycl -fsycl-device-only -Xclang -emit-llvm %s -o %t
+// RUN: sycl-post-link -split-esimd -lower-esimd -O0 -S %t -o %t.table
+// RUN: FileCheck %s -input-file=%t_esimd_0.ll --check-prefixes=CHECK
```
Comment (sarnex) [inline sycl/test/esimd/memory_properties.cpp:3]: Nit: I dont think you need `--check-prefixes=CHECK`, it should be the default

File: sycl/include/sycl/ext/intel/esimd/common.hpp [hunk 8]
```diff
@@ -352,6 +350,239 @@ template <__ESIMD_NS::atomic_op Op> constexpr int get_num_args() {
...
+
+template <cache_hint Hint> class cache_hint_wrap {
+  template <cache_hint...> struct is_one_of_t;
+  template <cache_hint Last>
+  struct is_one_of_t<Last>
+      : std::conditional_t<Last == Hint, std::true_type, std::false_type> {};
+  template <cache_hint Head, cache_hint... Tail>
+  struct is_one_of_t<Head, Tail...>
+      : std::conditional_t<Head == Hint, std::true_type, is_one_of_t<Tail...>> {
+  };
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/intel/esimd/common.hpp:530]: nit, can easily be ignored.

I'd prefer this to be written such that it could look closer to https://godbolt.org/z/zrT3Ya6qx, i.e., using `constexpr` function instead of metaprogramming.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp [hunk 9]
```diff
@@ -581,6 +581,77 @@ ESIMD_INLINE __ESIMD_NS::SurfaceIndex __esimd_get_surface_index(MemObjTy obj)
...
+    __ESIMD_DNS::vector_type_t<uintptr_t, N> addrs,
+    __ESIMD_DNS::vector_type_t<Ty, N * __ESIMD_DNS::to_int<VS>()> old_values =
+        0)
+#ifdef __SYCL_DEVICE_ONLY__
+    ;
+#else  // __SYCL_DEVICE_ONLY__
+{
+  __ESIMD_UNSUPPORTED_ON_HOST;
+}
+#endif // __SYCL_DEVICE_ONLY__
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:620]: nit, might go to separate PR (if at all)
```
#ifdef __SYCL_DEVICE_ONLY__
#define __ESIMD_INTRIN_DECL_END ;
#else
#define __ESIMD_INTRIN_DECL_END { __ESIMD_UNSUPPORTED_ON_HOST }
#endif

__ESIMD_INTRIN type name(args) __ESIMD_INTRIN_DECL_END

File: sycl/test-e2e/ESIMD/esimd_test_utils.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/esimd_test_utils.hpp:14]: I'll remove the other white-space (at L16) now. 

BTW, the define is probably not needed here (windows.h is not included here). It can be removed separately later.

File: sycl/test-e2e/ESIMD/unified_memory_api/block_load_usm.cpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+// The block_load() calls in this test do not use mask or cache-hint
+// properties to not impose using PVC features.
+
+#include "Inputs/block_load.hpp"
+
+int main() {
+  auto Q = queue{gpu_selector_v};
+  esimd_test::printTestLabel(Q);
+
+  constexpr bool TestPVCFeatures = true;
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/unified_memory_api/block_load_usm.cpp:22]: I did this intentional.

!TestPVCFeatures is supposed to be read as "Don't test-pvc-features"
and TestPVCFeatures in 'block_Load_usm_pvc.cpp' is read as "test-pvc-features'.

File: sycl/test/esimd/memory_properties.cpp (hunk 7 above)
Comment (v-klochkov) [reply sycl/test/esimd/memory_properties.cpp:3]: Yes, thank you. It was copied from other test that tested compilation in 2 modes with and without force-stateless-mem
I will remove `--check-prefixes=CHECK`.
We'll need `--check-prefixes' when start adding accessor-based API to this test.

File: sycl/include/sycl/ext/intel/esimd/common.hpp (hunk 8 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/common.hpp:530]: Thank you.
Here I only copied it from experimental. We have several/many like this. It deserves a special/separate PR.

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:597]: Will fix here and in one more place below. Thank you.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 11]
```diff
@@ -280,7 +283,7 @@ scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
 template <typename Tx, int N,
           typename Flags = overaligned_tag<detail::OperandSize::OWORD>>
 __ESIMD_API std::enable_if_t<is_simd_flag_type_v<Flags>, simd<Tx, N>>
-block_load(const Tx *addr, Flags = {}) {
+block_load(const Tx *addr, Flags) {
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:286]: It is not a breaking change. And this change is also required because now the defaultly chosen function for block_load(ptr) call must be not this one, but block_load(ptr, props={}).

File: sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp (hunk 9 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp:620]: Good idea, but yes - it is for separate PR sometime later.

File: llvm/lib/SYCLLowerIR/ESIMD/ESIMDVerifier.cpp [hunk 12]
```diff
@@ -60,6 +60,8 @@ static const char *LegalSYCLFunctions[] = {
     "^sycl::_V1::exp",
     "^sycl::_V1::bit_cast<.+>",
     "^sycl::_V1::operator.+<.+>",
+    "^sycl::_V1::ext::oneapi::experimental::properties",
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply llvm/lib/SYCLLowerIR/ESIMD/ESIMDVerifier.cpp:63]: Very good question. I did not realize that the need to add these 2 lines was signal about a potential problem.

I analyzed it a bit and as I can see it is the convenience payment for syntax sugar coming from 
a) convenient creation of 'properties' class using a number of properties following in random order
b) passing `properties` object to `block_load()` and having the default value for that argument:
```c++
esimd::properties props{ext::oneapi::experimental::alignment<16>};
auto v = block_load<float, N>(ptr, props);
```

The very first IR (after C++ FE, at ESIMDVerifier.cpp) has the real 'properties' object creation, and that object is then passed to `block_load()`
```
  %"class.sycl::_V1::ext::oneapi::experimental::properties" = type { %"class.std::tuple" }
  %"class.std::tuple" = type { i8 }
  ...

  %props2 = alloca %"class.sycl::_V1::ext::oneapi::experimental::properties", align 1
  %props2.ascast = addrspacecast ptr %props2 to ptr addrspace(4)
  ...
  call spir_func void @_ZN4sycl3_V13ext6oneapi12experimental10propertiesISt5tupleIJNS3_14property_valueINS3_13alignment_keyEJSt17integral_constantIiLi16EEEEEEEEC2IJSA_EEEDpT_(ptr addrspace(4) noundef align 1 dereferenceable_or_null(1) %props2.ascast, ptr noundef byval(%"struct.sycl::_V1::ext::oneapi::experimental::property_value") align 1 %agg.tmp.ascast.ascast) #12
    // This function calls "^sycl::_V1::ext::oneapi::experimental::detail::ExtractProperties" from ESIMDVerifier.cpp filter.
  ...
  call void @llvm.memcpy.p4.p4.i64(ptr addrspace(4) align 1 %agg.tmp2.ascast, ptr addrspace(4) align 1 %props2.ascast, i64 1, i1 false), !tbaa.struct !72
  %agg.tmp2.ascast.ascast = addrspacecast ptr addrspace(4) %agg.tmp2.ascast to ptr
  ...
  call spir_func void @_ZN4sycl3_V13ext5intel5esimd10block_loadIfLi32ENS1_6oneapi12experimental10propertiesISt5tupleIJNS6_14property_valueINS6_13alignment_keyEJSt17integral_constantIiLi16EEEEEEEEEEENSt9enable_ifIXsr3ext6oneapi12experimentalE18is_property_list_vIT1_EENS3_4simdIT_XT0_EEEE4typeEPKSJ_SH_(ptr addrspace(4) sret(%"class.sycl::_V1::ext::intel::esimd::simd") align 128 %va.ascast, ptr addrspace(4) noundef %add.ptr, ptr noundef byval(%"class.sycl::_V1::ext::oneapi::experimental::properties") align 1 %agg.tmp2.ascast.ascast) #12
```
Because that `properties` object is not used inside `block_load()`, the code related to 'properties' is removed later.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 13]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+      "Incorrect alignment for the data type");
+  static_assert(FDS == lsc_data_size::u16 || FDS == lsc_data_size::u8 ||
+                    FDS == lsc_data_size::u32 || FDS == lsc_data_size::u64,
+                "Conversion data types are not supported");
+  constexpr int SmallIntFactor32Bit =
+      (FDS == lsc_data_size::u16) ? 2 : (FDS == lsc_data_size::u8 ? 4 : 1);
+  static_assert(NElts > 0 && NElts % SmallIntFactor32Bit == 0,
+                "Number of elements is not supported by Transposed load");
+
+  constexpr bool Use64BitData =
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:495]: First I did some simplification/unification + brushed that code for 2 functions here c8a7f8388af16204b1ced6d2d265f6e341944386
Then when it was mostly ready for factoring, I decided to not doing that yet. The new function would be a bit odd: it would need to do lots of static_asserts + return `Use64BitData`.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 14]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+///                      or 256(only if alignment is 8-bytes or more);
+///     for 1-byte data: 4, 8, 12, 16, 32, 64, 128, 256,
+///                      or 512(only if alignment is 8-bytes or more).
+/// R3: The target device device must be PVC or newer.
+/// (R1), (R2) and (R2) are not applied if there are no cache hints.
+template <typename T, int N,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list<PropertyListT>::value,
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Yes, done in c8a7f8388af16204b1ced6d2d265f6e341944386

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 15]
```diff
@@ -367,6 +372,551 @@ block_store(Tx *addr, simd<Tx, N> vals, Flags = {}) {
...
+    simd<T, N>>
+block_load(const T *ptr, simd_mask<1> pred,
+           const PropertyListT &props =
+               ext::oneapi::experimental::detail::empty_properties_t{}) {
+  using L1HintProp =
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Taking into account that it is pretty common operation for property-based functions it is must have. Thank you!
Added in c8a7f8388af16204b1ced6d2d265f6e341944386

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in c8a7f8388af16204b1ced6d2d265f6e341944386

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in c8a7f8388af16204b1ced6d2d265f6e341944386

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in c8a7f8388af16204b1ced6d2d265f6e341944386

File: sycl/include/sycl/ext/intel/esimd/memory_properties.hpp [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/fpga_utils.hpp>
+#include <sycl/ext/oneapi/experimental/annotated_ptr/annotated_ptr.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+#include <sycl/ext/oneapi/properties/property_value.hpp>
+
+#define SYCL_EXT_INTEL_ESIMD_MEMORY_PROPERTIES 1
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory_properties.hpp:16]: good idea, thanks

File: sycl/test-e2e/ESIMD/unified_memory_api/block_load_usm.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/test-e2e/ESIMD/unified_memory_api/block_load_usm.cpp:22]: Agree to disagree, but not a big deal :)

**Review by sarnex — APPROVED** [review]
Looks really good, nice!

File: sycl/test/esimd/enums.cpp [hunk 17]
```diff
@@ -3,7 +3,7 @@
 // This test checks compilation of various ESIMD enum types. Those which are
 // deprecated must produce deprecation messages.
 
-#include <sycl/ext/intel/esimd/common.hpp>
+#include <sycl/ext/intel/esimd.hpp>
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/test/esimd/enums.cpp:6]: esimd.hpp is what users are supposed to include if need ESIMD feature.
This particular test included only esimd/common.hpp and had some non-reasonable assumptions on inclusion of API required for the test.
Thank you.
